# 💚Stepwise regression
  

# ✅backward stepwise regression
   

In [ ]:
#we here use a built-in dataset in R which is mtcars
head(mtcars)

mpg  cyl disp hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160  110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160  110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108   93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360  175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225  105 2.76 3.460 20.22 1  0  3    1

In [ ]:
model = lm(mpg ~ . ,data= mtcars)
formula(model)

mpg ~ cyl + disp + hp + drat + wt + qsec + vs + am + gear + carb

In [ ]:
summary(model)


Call:
lm(formula = mpg ~ ., data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.4506 -1.6044 -0.1196  1.2193  4.6271 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept) 12.30337   18.71788   0.657   0.5181  
cyl         -0.11144    1.04502  -0.107   0.9161  
disp         0.01334    0.01786   0.747   0.4635  
hp          -0.02148    0.02177  -0.987   0.3350  
drat         0.78711    1.63537   0.481   0.6353  
wt          -3.71530    1.89441  -1.961   0.0633 .
qsec         0.82104    0.73084   1.123   0.2739  
vs           0.31776    2.10451   0.151   0.8814  
am           2.52023    2.05665   1.225   0.2340  
gear         0.65541    1.49326   0.439   0.6652  
carb        -0.19942    0.82875  -0.241   0.8122  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.65 on 21 degrees of freedom
Multiple R-squared:  0.869,	Adjusted R-squared:  0.8066 
F-statistic: 13.93 on 10 and 21 DF,  p-value: 3.793e-07

In [ ]:
step(model, direction = "backward")


Start:  AIC=70.9
mpg ~ cyl + disp + hp + drat + wt + qsec + vs + am + gear + carb

       Df Sum of Sq    RSS    AIC
- cyl   1    0.0799 147.57 68.915
- vs    1    0.1601 147.66 68.932
- carb  1    0.4067 147.90 68.986
- gear  1    1.3531 148.85 69.190
- drat  1    1.6270 149.12 69.249
- disp  1    3.9167 151.41 69.736
- hp    1    6.8399 154.33 70.348
- qsec  1    8.8641 156.36 70.765
<none>              147.49 70.898
- am    1   10.5467 158.04 71.108
- wt    1   27.0144 174.51 74.280

Step:  AIC=68.92
mpg ~ disp + hp + drat + wt + qsec + vs + am + gear + carb

       Df Sum of Sq    RSS    AIC
- vs    1    0.2685 147.84 66.973
- carb  1    0.5201 148.09 67.028
- gear  1    1.8211 149.40 67.308
- drat  1    1.9826 149.56 67.342
- disp  1    3.9009 151.47 67.750
- hp    1    7.3632 154.94 68.473
<none>              147.57 68.915
- qsec  1   10.0933 157.67 69.032
- am    1   11.8359 159.41 69.384
- wt    1   27.0280 174.60 72.297

Step:  AIC=66.97
mpg ~ disp + hp + drat + wt + qsec + am


Call:
lm(formula = mpg ~ wt + qsec + am, data = mtcars)

Coefficients:
(Intercept)           wt         qsec           am  
      9.618       -3.917        1.226        2.936  


In [ ]:
mod=lm(formula = mpg ~ wt + qsec + am, data = mtcars)
summary(mod)


Call:
lm(formula = mpg ~ wt + qsec + am, data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.4811 -1.5555 -0.7257  1.4110  4.6610 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   9.6178     6.9596   1.382 0.177915    
wt           -3.9165     0.7112  -5.507 6.95e-06 ***
qsec          1.2259     0.2887   4.247 0.000216 ***
am            2.9358     1.4109   2.081 0.046716 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.459 on 28 degrees of freedom
Multiple R-squared:  0.8497,	Adjusted R-squared:  0.8336 
F-statistic: 52.75 on 3 and 28 DF,  p-value: 1.21e-11


❤Remember: lower AIC is better (:
in backward case we obtain an AIC = 61.31 and the most predictor variables are  wt , qsec and am

# ✅Forward stepwise regression

In [ ]:
initial_model = lm(mpg ~ 1 ,data= mtcars)
summary(initial_model)


Call:
lm(formula = mpg ~ 1, data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-9.6906 -4.6656 -0.8906  2.7094 13.8094 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   20.091      1.065   18.86   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 6.027 on 31 degrees of freedom


In [ ]:
step(initial_model, direction = "forward")

Start:  AIC=115.94
mpg ~ 1




Call:
lm(formula = mpg ~ 1, data = mtcars)

Coefficients:
(Intercept)  
      20.09  


In [ ]:
step(initial_model, direction = "forward",scope=formula(model))

Start:  AIC=115.94
mpg ~ 1

       Df Sum of Sq     RSS     AIC
+ wt    1    847.73  278.32  73.217
+ cyl   1    817.71  308.33  76.494
+ disp  1    808.89  317.16  77.397
+ hp    1    678.37  447.67  88.427
+ drat  1    522.48  603.57  97.988
+ vs    1    496.53  629.52  99.335
+ am    1    405.15  720.90 103.672
+ carb  1    341.78  784.27 106.369
+ gear  1    259.75  866.30 109.552
+ qsec  1    197.39  928.66 111.776
<none>              1126.05 115.943

Step:  AIC=73.22
mpg ~ wt

       Df Sum of Sq    RSS    AIC
+ cyl   1    87.150 191.17 63.198
+ hp    1    83.274 195.05 63.840
+ qsec  1    82.858 195.46 63.908
+ vs    1    54.228 224.09 68.283
+ carb  1    44.602 233.72 69.628
+ disp  1    31.639 246.68 71.356
<none>              278.32 73.217
+ drat  1     9.081 269.24 74.156
+ gear  1     1.137 277.19 75.086
+ am    1     0.002 278.32 75.217

Step:  AIC=63.2
mpg ~ wt + cyl

       Df Sum of Sq    RSS    AIC
+ hp    1   14.5514 176.62 62.665
+ carb  1   13.7724 177.40 62.805
<no


Call:
lm(formula = mpg ~ wt + cyl + hp, data = mtcars)

Coefficients:
(Intercept)           wt          cyl           hp  
   38.75179     -3.16697     -0.94162     -0.01804  


Mais dans ce cas, la sélection directe(forward) nous avons obtenu l’AIC= 62.66 et les variables les plus significatives sont : wt, cyl et hp POURQUOI ? dans ce cas, la sélection en arrière est préférable car elle donne un AIC inférieur à l’avant

❤Le choix entre une régression progressive en avant et en arrière dépend souvent de la question de recherche et des objectifs spécifiques de l’analyse. La régression progressive vers l’avant peut être préférable lorsque le chercheur veut construire un modèle à partir de zéro, tandis que la régression progressive vers l’arrière peut être préférable lorsque le chercheur veut tester un modèle complet et identifier les variables les plus importantes pour la prédiction.

# ***AIC***
 L'Information de Critère d'Akaike (AIC) est une mesure de la qualité d'ajustement d'un modèle statistique, souvent utilisée pour la sélection de modèles.

Le AIC est déterminé en calculant -2 x log de la fonction de vraisemblance maximale du modèle, ajusté pour la complexité du modèle. Plus précisément, le AIC est défini comme suit:

AIC = -2 log(L) + 2k

où L est la fonction de vraisemblance maximale du modèle et k est le nombre de paramètres du modèle.

Le AIC favorise les modèles qui ont une bonne qualité d'ajustement (c'est-à-dire une faible somme des carrés des résidus) tout en étant simples (c'est-à-dire ayant moins de paramètres). Par conséquent, un modèle avec un AIC plus faible est considéré comme meilleur qu'un modèle avec un AIC plus élevé.

Il est courant d'utiliser le AIC pour comparer plusieurs modèles différents et choisir celui qui a le AIC le plus faible comme modèle final. Cependant, il est important de noter que le AIC ne doit pas être utilisé seul et doit être utilisé en conjonction avec d'autres critères de sélection de modèle.

  La vraisemblance (likelihood en anglais) est une mesure de la probabilité d'observer une certaine série de données, étant donné un modèle statistique avec des paramètres estimés. En d'autres termes, la vraisemblance est la probabilité de voir les données que nous avons observées, si nous supposons que notre modèle est correct et que les paramètres que nous avons estimés sont les vraies valeurs des paramètres.

  Le but de l'inférence statistique est de trouver la meilleure estimation des paramètres d'un modèle donné, et la vraisemblance est souvent utilisée comme mesure d'évaluation de la qualité des estimations de paramètres. La fonction de vraisemblance est souvent utilisée pour trouver les estimations de maximum de vraisemblance (Maximum Likelihood Estimation ou MLE en anglais), qui est la méthode d'estimation des paramètres la plus courante dans de nombreux modèles statistiques.
   